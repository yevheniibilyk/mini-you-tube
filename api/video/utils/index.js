const mime = require('mime-types');
const { apiUrl } = require('../../../src/config');

const mediaStaticUrl = `${apiUrl}/media`;

const isVideoFile = (filePath) => {
  const mimeType = mime.lookup(filePath);

  return mimeType.startsWith('video/');
};

const formVideoResponse = (fileName) => ({
  name: fileName,
  url: `${mediaStaticUrl}/${fileName}`,
  thumbnailUrl: `${mediaStaticUrl}/${fileName}_THUMBNAIL.png`,
});

module.exports = {
  isVideoFile,
  formVideoResponse,
};
