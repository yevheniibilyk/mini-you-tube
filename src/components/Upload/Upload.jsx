import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DropzoneArea } from 'material-ui-dropzone';
import { CircularProgress } from '@material-ui/core';
import { withSnackbar } from 'notistack';
import { axios } from '../../utils';
import './styles/upload-page.scss';

// 10MB
const MAX_FILE_SIZE = 20 * 1048576;

class Upload extends Component {
  constructor(props){
    super(props);

    this.state = {
      loading: false
    };
  }

  onDropRejected = (alertText, variant) => {
    const { enqueueSnackbar } = this.props;

    if (!["error", "info"].includes(variant)) {
      return;
    }

    enqueueSnackbar(alertText,  { variant });
  }

  uploadFiles = async (files) => {
    if (!files || !files.length) {
      return;
    }

    this.setState({ loading: true });

    await Promise.all(
      files.map(this.uploadFile)
    );

    this.setState({ loading: false });
  }

  uploadFile = async (file) => {
    const { enqueueSnackbar } = this.props;

    const data = new FormData();

    data.append('file', file);

    try {
      await axios.post('/api/upload', data);
      enqueueSnackbar(`File ${file.name} uploaded`, { variant: 'success' });
    } catch (e) {
      enqueueSnackbar(`Upload failed for ${file.name}`,  { variant: 'error' })
    }
  }

  render() {
    const { loading } = this.state;

    return (
      <div className="upload-page">
        <div className="upload-page__dropzone">
          {!loading ? (
            <DropzoneArea
              onChange={this.uploadFiles}
              onAlert={this.onDropRejected}
              acceptedFiles={['video/*']}
              maxFileSize={MAX_FILE_SIZE}
              filesLimit={10}
              disabled={loading}
              showAlerts={false}
              showPreviews={false}
              showPreviewsInDropzone={false}
            />
          ) : (<CircularProgress className="upload-page--loading" size={100}/>)}
        </div>
      </div>
    )
  }
}

Upload.propTypes = {
  enqueueSnackbar: PropTypes.func.isRequired
}

export default withSnackbar(Upload);