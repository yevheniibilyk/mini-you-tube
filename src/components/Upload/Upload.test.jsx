import React from 'react';
import renderer from 'react-test-renderer';
import { axios } from '../../utils';
import Upload from './Upload';

jest.mock('../../utils/axios');
jest.mock('notistack', () => ({
  withSnackbar(component) {
    component.defaultProps = {
      ...component.defaultProps,
      enqueueSnackbar: jest.fn(),
    };
    return component;
  },
}));

const renderComponent = () => renderer.create(<Upload />);

describe('Upload page', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  test('render Upload page', () => {
    const component = renderComponent();

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should call enqueueSnackbar on dropzone error and info alert', () => {
    const component = renderComponent();
    const instance = component.getInstance();

    instance.onDropRejected('Success', 'success');
    expect(instance.props.enqueueSnackbar).not.toHaveBeenCalled();

    instance.onDropRejected('Info msg', 'info');
    expect(instance.props.enqueueSnackbar).toHaveBeenCalledWith('Info msg', { variant: 'info' });

    instance.onDropRejected('Error msg', 'error');
    expect(instance.props.enqueueSnackbar).toHaveBeenCalledWith('Error msg', { variant: 'error' });
  });

  test('should not start files uploading if files list empty', () => {
    const component = renderComponent();
    const instance = component.getInstance();

    instance.uploadFiles();

    expect(instance.props.enqueueSnackbar).not.toHaveBeenCalled();
  });

  test('should upload files', async () => {
    axios.post.mockImplementationOnce(() => Promise.resolve({}));

    const component = renderComponent();

    const instance = component.getInstance();
    const filename = 'video.mp4';

    await instance.uploadFiles([{ name: filename }]);

    expect(instance.props.enqueueSnackbar)
      .toHaveBeenCalledWith(`File ${filename} uploaded`, { variant: 'success' });
  });

  test('should handle failed file upload', async () => {
    axios.post.mockImplementationOnce(() => Promise.reject());

    const component = renderComponent();
    const instance = component.getInstance();

    const filename = 'video.mp4';

    await instance.uploadFiles([{ name: filename, withError: true }]);

    expect(instance.props.enqueueSnackbar)
      .toHaveBeenCalledWith(`Upload failed for ${filename}`, { variant: 'error' });
  });
});
