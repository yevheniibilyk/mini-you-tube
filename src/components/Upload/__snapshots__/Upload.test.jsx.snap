// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload page render Upload page 1`] = `
<div
  className="upload-page"
>
  <div
    className="upload-page__dropzone"
  >
    <div
      className="MuiDropzoneArea-root"
      onBlur={[Function]}
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        accept="video/*"
        autoComplete="off"
        multiple={true}
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        tabIndex={-1}
        type="file"
      />
      <div
        className="MuiDropzoneArea-textContainer"
      >
        <p
          className="MuiTypography-root MuiDropzoneArea-text MuiTypography-h5"
        >
          Drag and drop a file here or click
        </p>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiDropzoneArea-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
